--- package/system/opkg/files/customfeeds.conf
+++ package/system/opkg/files/customfeeds.conf
@@ -1,3 +1,10 @@
 # add your custom package feeds here
 #
 # src/gz example_feed_name http://www.example.com/path/to/files
+
+src/gz unwired_one_base http://files.unwireddevices.com/files/openwrt/ccalmer/1.0/packages/base
+src/gz unwired_one_luci http://files.unwireddevices.com/files/openwrt/ccalmer/1.0/packages/luci
+src/gz unwired_one_mgmt http://files.unwireddevices.com/files/openwrt/ccalmer/1.0/packages/management
+src/gz unwired_one_pkgs http://files.unwireddevices.com/files/openwrt/ccalmer/1.0/packages/packages
+src/gz unwired_one_rout http://files.unwireddevices.com/files/openwrt/ccalmer/1.0/packages/routing
+src/gz unwired_one_tele http://files.unwireddevices.com/files/openwrt/ccalmer/1.0/packages/telephony
