Index: config/Config-kernel.in
=====================================
--- config/Config-kernel.in
+++ config/Config-kernel.in
@@ -17,6 +17,12 @@ config KERNEL_SWAP
 	bool "Support for paging of anonymous memory (swap)"
 	default y
 
+config KERNEL_VERMAGIC
+	string "Specify vermagic for the build"
+	default ""
+	help
+	  Specify kernel vermagic. Leave it empty if you don't know what it is.
+
 config KERNEL_DEBUG_FS
 	bool "Compile the kernel with debug filesystem enabled"
 	default y
Index: include/kernel-defaults.mk
=====================================
--- include/kernel-defaults.mk
+++ include/kernel-defaults.mk
@@ -112,7 +112,9 @@
 	$(call Kernel/SetNoInitramfs)
 	rm -rf $(KERNEL_BUILD_DIR)/modules
 	$(_SINGLE) [ -d $(LINUX_DIR)/user_headers ] || $(MAKE) $(KERNEL_MAKEOPTS) INSTALL_HDR_PATH=$(LINUX_DIR)/user_headers headers_install
-	$(SH_FUNC) grep '=[ym]' $(LINUX_DIR)/.config | LC_ALL=C sort | md5s > $(LINUX_DIR)/.vermagic
+	VERMAGIC=`grep CONFIG_KERNEL_VERMAGIC $(TOPDIR)/.config | awk -F= '{print $$$$2}' | sed -e 's/"\(.*\)"/\1/g'`; \
+	[ -n "$$$$VERMAGIC" ] && echo $$$$VERMAGIC > $(LINUX_DIR)/.vermagic || \
+	( $(SH_FUNC) grep '=[ym]' $(LINUX_DIR)/.config | LC_ALL=C sort | md5s > $(LINUX_DIR)/.vermagic )
 endef
 
 define Kernel/Configure/Initramfs
